### PR TITLE
WIP: AbstractVariationalState

### PR DESCRIPTION
**Pull request by gcarleo**
This is to start a discussion on a concrete possible interface for ```AbstractVariationalState``` 

 Essentially, this encapsulates a quantum state (either pure or mixed) depending on a set of parameters, and that supports operations such as computing quantum expectation values and their gradients.

Notice that this is broader concept than ```AbstractMachine```, since it allows also computing expectation values. 

This is somehow in the spirit of #404 but a bit more general, so that this could be used in all applications we have. Notice that specific implementations of this relying on MCMC might want to implement a caching mechanism, such that samples to compute expectation values are not recomputed unless variational parameters in the state are changed.  

Other families of states can also derive from this interface, and implement their own strategy (possibly non-stochastic) to compute expectation values and gradients.

### CODE REVIEW

**Comment by gcarleo** [comment]
I think that introducing this abstraction is useful also to have a more consistent hierarchy of objects in netket and resolves some existing ambiguity with the concept of ```Machine```.

Essentially, 
1. A ```Machine``` is just a parameterized high dimensional function. It has no intrinsic notion of quantum mechanics. 

2. A ```VariationalState``` is a parameterized quantum state, representing pure or mixed states, and has the properties expected from a quantum state, including the ability to compute expectation values and quantum geometric tensors. Specific implementation can internally use ```Machine``` and ```Sampler```, others can use other approaches. 

3. High-order applications work exclusively with ```VariationalState``` not with ```Machine```

File: netket/variational/abstract_state.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            In the case of a pure state, samples x=(x_1,..,x_N) are returned according to the
+            Born probability P(x)=|Psi(x)|^2. In the case of mixed states, P(x)=Rho(x,x).
+
+        Args:
+            n_samples (int): the number of samples required.
+
+        Returns:
+            array: An array of shape (n_samples,N) containing samples x.
+        """
+        raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/variational/abstract_state.py:73]: I had suggested adding this with the rationale that you don't necessarily need to fully implement the State interface in every implementations.

There are some use cases where one might want to write other drivers or functionality where he needs to access the samples explicitly, modify them, and then compute expectation values on the modified states (computing the reny-2 entropy, for example).

Implementing `samples` allows this.

**Comment by PhilipVinc** [comment]
> There is always an overhead to caching (with respect to performance, but especially with respect to understanding the flow of the code).

I agree, but `hiding` something like sampling as a well documented implementation detail of some precise variational state is not necessarily bad design.
We have to construct a well thought out interface, though.

I have two remarks:

```python
for _ in range(vmc_steps):
    vmcstate = VmcState(machine, sampler, n_samples=10000)
    energy, grad = vmcstate.mean_and_grad(hamiltonian)
   ....
    machine.parameters += dp
```
1) As you say code duplication usually signals that something is wrong. Now imagine I want to have a different state, called `ExactMachineState` that performs calculations on the whole Hilbert space. Since you are creating an explicit vmcstate in the optimization loop with the information of the samples this loop won't generalise well.

I think If the code you posted above is the implementation of the driver, you are leaking details of the implementation (how vmc state computes expectation values) to the outside world.

I don't necessarily dislike your idea of an immutable vmcstate (or, rather, `VmcSampledState`), though.

2)
>     # Changing the machine does not magically change VmcState, we just create a new one
I think the issue right now with our interface is that it retains references a bit everywhere.
In Giuseppe's implementation I can do
```
machine = Rbmspin(...)
sampler = MetropolisLocal(machine)
vmcstate = VmcState(machine, sampler, n_samples=1000) # redundant to pass machine twice but.. ok.
Eave1 = vmcstate.expval(H)
machine.parameters = [1,2,3,4...]
Eave2 = vmcstate.expval(H)

```
and the latter expectation value will be computed with the samples of the old weights, but with the values of the new one.

A more functional approach, quite easy, would be to store in `vmcstate` a copy of the machine, so that is you modify the original the one inside is not affected, and then expose a `get_parameters` `set_parameters` interface through the state itself in order to be able to detect when weights are modified and therefore do `self._samples = None` or something similar.

The problem is that this is a bit incompatible with our current mental model: right now there is only one machine, that we copy everywhere. In a functional paradigm most functions would get their copy of the machine, but when you optimise it you must return a new one.

This approach would not be a big problem when using things like Jax or PyTorch, because their programming paradigm is functional already.

**Comment by gcarleo** [comment]
Thanks @femtobit and @PhilipVinc for these comments, indeed it is a bit tricky to come up with an interface that satisfies all the constraints... 

For the moment I just wanted to understand what happens in @femtobit 's approach when you want to compute expectation values of other observables, in the training loop. If I understand your proposal, the samples would be generated only at construction level of ```VmcState``` and then reused when computing expectation values, right? 

Concerning the other issue (1) raised by @PhilipVinc : 
This is a bit tricky indeed. What I wanted from this design, is to remove all details on how expectation values are computed in the high-level driver, thus retaining the sampler in the driver should not be allowed. In this sense, I couldn't come up with ah interface that satisfies this requirement and that also allows to keep the state of the sampler (so that at each new iteration you can, in principle, start from the previous state of the MCMC, for example).  

If am not mistaken, the only option is to construct a ```VmcState``` outside of the high-level driver? Using the functional approach might help to avoid the issue raised by @femtobit, but that's not clear yet how (to me). 

On a related note, I like the idea of not duplicating the parameterization logic that is already in ```Machine```.

**Comment by gcarleo** [comment]
I think it's time to get back to this and finalize this interface, which is really needed... 
I think we all agree on the fact that the general abstract interface is just something that allows to compute amplitudes, expectation values etc and other operations that can be done on quantum states like measurements (generating samples). 

Now, the issue is what are we going to do with the methods ```gradient``` and ```parameters```. Since this is an abstract class for a variational state, by definition this is something parameterized, so it seems natural to me that it returns its parameters? 

This in practice just means that we will define ```JaxMachine``` or something to be derived from this abstract variational state and it is constructed with a jax model and a sampler. 

Now, the trickier part is about the gradient of the expectation value. Do we want to make this part of the interface of the variational class or not? The main issue is that there are other applications where you don't want the gradient of expectation value but rather the gradient of the amplitudes or something... @femtobit @PhilipVinc ?

**Comment by PhilipVinc** [comment]
About the comment on `parameters` @femtobit had raised I'd say that that should be a pytree. When #517 goes through we really will have no more reason to have a contiguous representation of the weights.

**Comment by PhilipVinc** [comment]
As for the gradient: There is little difference if the code to do the gradient lives here or somewhere else, isn't it? In the end, we need this code somewhere. The advantage of having it within the machines is that we get single-dispatch automatically based on the type of the machine

File: netket/variational/abstract_state.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+    @property
+    @abc.abstractmethod
+    def parameters(self):
+        r"""Parameters of this variational state"""
+        raise NotImplementedError
+
```
[the remark this answers is not in the document]
Comment (gcarleo) [reply netket/variational/abstract_state.py:80]: Addressed by @PhilipVinc , indeed it makes sense now to have pytrees everywhere

File: netket/variational/abstract_state.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gcarleo) [reply netket/variational/abstract_state.py:73]: yes I'd also leave this method as it is, in the sense it getting samples from some quantum state is an operation that is well defined for all class of wave functions, even if it's not neural nets. In the case of other states (say for exact states or something) the sampling will be specialized and does not need to rely on our standard samplers, for example...

File: netket/variational/abstract_state.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                               Notice that for most variational states it is
+                               exponentially expensive in N to compute the normalization.
+
+            Returns:
+                array: An array of shape (batch_size,n_par) containing the gradient of the log amplitudes.
+        """
+        raise NotImplementedError
+
+    @property
+    def hilbert(self, number):
```
[the remark this answers is not in the document]
Comment (gcarleo) [reply netket/variational/abstract_state.py:124]: yeah thanks that was something weird

**Comment by PhilipVinc** [comment]
Also this, implemented with #539
